src/project/manifest/pyproject.rs

```diff
@@ -147,46 +149,77 @@ fn version_or_url_to_nameless_matchspec(
+                } else {
+                    Ok(PyProjectToml { inner: pyproject })
                 }
             }
-            environments.insert(extra.clone(), features);
         }
     }
-    environments
-}
 
-/// Parses a non-pixi pyproject.toml string.
-pub fn pyproject(source: &str) -> Result<PyProjectToml, Report> {
-    match toml_edit::de::from_str::<PyProjectToml>(source).map_err(TomlError::from) {
-        Err(e) => e.to_fancy("pyproject.toml", source),
-        Ok(pyproject) => {
-            // Make sure [project] exists in pyproject.toml,
-            // This will ensure project.name is defined
-            if pyproject.project.is_none() {
-                TomlError::NoProjectTable.to_fancy("pyproject.toml", source)
-            } else {
-                Ok(pyproject)
+    pub fn name(&self) -> String {
+        self.project().name.clone()
+    }
+
+    pub fn project(&self) -> &Project {
+        self.inner.project.as_ref().unwrap()
+    }
+
+    /// Builds a list of pixi environments from pyproject groups of extra dependencies:
+    ///  - one environment is created per group of extra, with the same name as the group of extra
+    ///  - each environment includes the feature of the same name as the group of extra
+    ///  - it will also include other features inferred from any self references to other groups of extras
+    pub fn environments_from_extras(&self) -> HashMap<String, Vec<String>> {
+        let mut environments = HashMap::new();
+        if let Some(extras) = &self.project().optional_dependencies {
+            let pname = pep508_rs::PackageName::new(self.name()).unwrap();
+            for (extra, reqs) in extras {
+                let mut features = vec![extra.to_string()];
+                // Add any references to other groups of extra dependencies
+                for req in reqs.iter() {
+                    if pname == req.name {
+                        for extra in &req.extras {
+                            features.push(extra.to_string())
+                        }
+                    }
+                }
+                environments.insert(extra.clone(), features);
             }
         }
+        environments
+    }
+
+    /// Checks whether a path is a valid `pyproject.toml` for use with pixi by checking if it
+    /// contains a `[tool.pixi.project]` table.
+    pub fn is_pixi(path: &PathBuf) -> bool {
+        let source = fs::read_to_string(path).unwrap();
+        Self::is_pixi_str(&source).unwrap_or(false)
+    }
+    /// Checks whether a string is a valid `pyproject.toml` for use with pixi by checking if it
+    /// contains a `[tool.pixi.project]` table.
+    pub fn is_pixi_str(source: &str) -> Result<bool, Report> {
+        match source.parse::<DocumentMut>().map_err(TomlError::from) {
+            Err(e) => e.to_fancy("pyproject.toml", source),
+            Ok(doc) => Ok(doc["tool"]["pixi"]["project"].is_table()),
```

Comment:
I panic on this line if I init on this pyproject.toml:
```toml
[project]
name = "pyproject_demo"
requires-python = ">=3.9"
dependencies = [
    "rich",
]
optional-dependencies = { "test" = ["pytest"], "dev" = ["black", "flake8"], "doc" = ["mkdocs", "mkdocs-material"]}
version = "0.1.0"

[build-system]
requires = ["setuptools", "wheel"]
build-backend = "setuptools.build_meta"
```
on this error:
```
thread 'main' panicked at src/project/manifest/pyproject.rs:221:30:
index not found
```
I guess you should use the `.get` functionality here to get better errors..

Reply:
Oopsie. looks like I took the 'if it compiles it'll work' a bit too seriously :-/